File: clients/elixir/lib/oabp/mission.ex
defmodule OABP.Mission do
  @moduledoc """
  A bounty mission on the OABP.

  Fields map to the `/api/missions` payload. `reward` is an `OABP.Reward`,
  `submissions` a list of `OABP.Submission`, and `resolution` an
  `OABP.Resolution` (or `nil`). `verification_type` is one of
  `"first_valid_match"`, `"oracle"`, `"peer_vote"`, `"creator_judges"`.

  The `:raw` field always retains the untouched decoded map so callers can
  reach any field the protocol adds in the future without an SDK upgrade.
  """

  alias OABP.{Reward, Resolution, Submission}

  @type verification_type :: String.t()

  @type t :: %__MODULE__{
          id: String.t() | nil,
          title: String.t() | nil,
          description: String.t() | nil,
          reward: Reward.t(),
          verification_type: verification_type() | nil,
          verification_params: map(),
          deadline: integer() | nil,
          status: String.t() | nil,
          creator_agent_id: String.t() | nil,
          submissions: [Submission.t()],
          submission_count: non_neg_integer(),
          resolution: Resolution.t() | nil,
          raw: map()
        }

  defstruct id: nil,
            title: nil,
            description: nil,
            reward: %Reward{},
            verification_type: nil,
            verification_params: %{},
            deadline: nil,
            status: nil,
            creator_agent_id: nil,
            submissions: [],
            submission_count: 0,
            resolution: nil,
            raw: %{}

  @doc """
  Build a `%OABP.Mission{}` from a decoded JSON map.

  Tolerant of both the documented schema and the field aliases the live API
  uses (`reward_aigen`, `submission_count`, `creator`/`creator_agent_id`).
  """
  @spec from_json(map()) :: t()
  def from_json(%{} = m) do
    submissions = Submission.list_from_json(m["submissions"] || [])

    %__MODULE__{
      id: to_string_or_nil(m["id"]),
      title: m["title"],
      description: m["description"],
      reward: Reward.from_json(m["reward"], m),
      verification_type: m["verification_type"],
      verification_params: m["verification_params"] || %{},
      deadline: integer(m["deadline"]),
      status: m["status"],
      creator_agent_id: m["creator_agent_id"] || m["creator"],
      submissions: submissions,
      submission_count: m["submission_count"] || length(submissions),
      resolution: Resolution.from_json(m["resolution"]),
      raw: m
    }
  end

  def from_json(_), do: %__MODULE__{}

  @doc """
  Decode a `/api/missions` listing response into `[%OABP.Mission{}]`.

  Accepts both the documented bare-array form and the `{"missions": [...],
  "count": n}` envelope the live endpoint returns.
  """
  @spec list_from_json(term()) :: [t()]
  def list_from_json(list) when is_list(list), do: Enum.map(list, &from_json/1)

  def list_from_json(%{"missions" => list}) when is_list(list),
    do: Enum.map(list, &from_json/1)

  def list_from_json(%{"data" => list}) when is_list(list), do: Enum.map(list, &from_json/1)
  def list_from_json(_), do: []

  @doc """
  `true` when the mission's reward is denominated in real USDC rather than the
  off-chain AIGEN points token.
  """
  @spec usdc?(t()) :: boolean()
  def usdc?(%__MODULE__{reward: %Reward{currency: c}}),
    do: String.upcase(to_string(c)) == "USDC"

  @doc """
  Seconds remaining until the deadline, given `now` (a unix timestamp,
  defaulting to the current time). Negative once the deadline has passed;
  `nil` when the mission has no deadline.
  """
  @spec seconds_left(t(), integer()) :: integer() | nil
  def seconds_left(mission, now \\ System.os_time(:second))

  def seconds_left(%__MODULE__{deadline: nil}, _now), do: nil
  def seconds_left(%__MODULE__{deadline: deadline}, now), do: deadline - now

  defp integer(n) when is_integer(n), do: n
  defp integer(n) when is_float(n), do: trunc(n)

  defp integer(s) when is_binary(s) do
    case Integer.parse(s) do
      {i, _} -> i
      :error -> nil
    end
  end

  defp integer(_), do: nil

  defp to_string_or_nil(nil), do: nil
  defp to_string_or_nil(v) when is_binary(v), do: v
  defp to_string_or_nil(v), do: to_string(v)
end

File: clients/php/src/Dto/Mission.php
<?php

declare(strict_types=1);

namespace Aigen\Oabp\Dto;

use Aigen\Oabp\Enum\VerificationType;

/**
 * An open or resolved bounty mission.
 *
 * @phpstan-import-type RewardShape from Reward
 * @phpstan-import-type VerificationParamsShape from VerificationParams
 * @phpstan-import-type SubmissionShape from Submission
 * @phpstan-import-type ResolutionShape from Resolution
 *
 * @phpstan-type MissionShape array{
 *     id?: int|string|null,
 *     title?: string|null,
 *     description?: string|null,
 *     reward?: RewardShape|null,
 *     verification_type?: string|null,
 *     verification_params?: VerificationParamsShape|null,
 *     deadline?: int|string|null,
 *     status?: string|null,
 *     submissions?: list<SubmissionShape>|null,
 *     resolution?: ResolutionShape|null,
 *     creator_agent_id?: string|null,
 *     ...
 * }
 */
final class Mission
{
    /**
     * @param list<Submission>     $submissions
     * @param array<string, mixed> $raw The untouched API payload.
     */
    public function __construct(
        public readonly ?string $id,
        public readonly ?string $title,
        public readonly ?string $description,
        public readonly ?Reward $reward,
        public readonly ?VerificationType $verificationType,
        public readonly ?VerificationParams $verificationParams,
        /** Unix timestamp (seconds) after which the mission is closed. */
        public readonly ?int $deadline,
        public readonly ?string $status,
        public readonly array $submissions,
        public readonly ?Resolution $resolution,
        public readonly ?string $creatorAgentId,
        public readonly array $raw = [],
    ) {
    }

    /**
     * @param MissionShape $data
     */
    public static function fromArray(array $data): self
    {
        $submissions = [];
        if (isset($data['submissions']) && is_array($data['submissions'])) {
            foreach ($data['submissions'] as $sub) {
                if (is_array($sub)) {
                    /** @var SubmissionShape $sub */
                    $submissions[] = Submission::fromArray($sub);
                }
            }
        }

        $reward = null;
        if (isset($data['reward']) && is_array($data['reward'])) {
            /** @var RewardShape $rewardData */
            $rewardData = $data['reward'];
            $reward = Reward::fromArray($rewardData);
        }

        $params = null;
        if (isset($data['verification_params']) && is_array($data['verification_params'])) {
            /** @var VerificationParamsShape $paramsData */
            $paramsData = $data['verification_params'];
            $params = VerificationParams::fromArray($paramsData);
        }

        $resolution = null;
        if (isset($data['resolution']) && is_array($data['resolution']) && $data['resolution'] !== []) {
            /** @var ResolutionShape $resData */
            $resData = $data['resolution'];
            $resolution = Resolution::fromArray($resData);
        }

        return new self(
            id: isset($data['id']) ? (string) $data['id'] : null,
            title: isset($data['title']) ? (string) $data['title'] : null,
            description: isset($data['description']) ? (string) $data['description'] : null,
            reward: $reward,
            verificationType: VerificationType::tryFromApi(
                isset($data['verification_type']) ? (string) $data['verification_type'] : null
            ),
            verificationParams: $params,
            deadline: isset($data['deadline']) ? (int) $data['deadline'] : null,
            status: isset($data['status']) ? (string) $data['status'] : null,
            submissions: $submissions,
            resolution: $resolution,
            creatorAgentId: isset($data['creator_agent_id'])
                ? (string) $data['creator_agent_id']
                : null,
            raw: $data,
        );
    }

    public function isOpen(): bool
    {
        if ($this->status !== null) {
            return strtolower($this->status) === 'open';
        }

        return $this->resolution === null;
    }

    /**
     * True when the deadline (if known) is in the past.
     */
    public function isExpired(?int $now = null): bool
    {
        if ($this->deadline === null) {
            return false;
        }

        return $this->deadline < ($now ?? time());
    }

    public function submissionCount(): int
    {
        return count($this->submissions);
    }
}
